pkg/services/provisioning/dashboards/dashboard.go

```diff
@@ -3,44 +3,79 @@ package dashboards
 import (
 	"context"
 	"fmt"
-
 	"github.com/grafana/grafana/pkg/log"
+	"github.com/pkg/errors"
 )
 
-type DashboardProvisioner struct {
-	cfgReader *configReader
-	log       log.Logger
+type DashboardProvisioner interface {
+	Provision() error
+	PollChanges(ctx context.Context)
+}
+
+type DashboardProvisionerImpl struct {
+	log         log.Logger
+	fileReaders []*fileReader
 }
 
-func NewDashboardProvisioner(configDirectory string) *DashboardProvisioner {
-	log := log.New("provisioning.dashboard")
-	d := &DashboardProvisioner{
-		cfgReader: &configReader{path: configDirectory, log: log},
-		log:       log,
+type DashboardProvisionerFactory func(string) (DashboardProvisioner, error)
+
+func NewDashboardProvisionerImpl(configDirectory string) (*DashboardProvisionerImpl, error) {
+	logger := log.New("provisioning.dashboard")
+	cfgReader := &configReader{path: configDirectory, log: logger}
+	configs, err := cfgReader.readConfig()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to read dashboards config")
 	}
 
-	return d
-}
+	fileReaders, err := getFileReaders(configs, logger)
 
-func (provider *DashboardProvisioner) Provision(ctx context.Context) error {
-	cfgs, err := provider.cfgReader.readConfig()
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "Failed to initialize file readers")
+	}
+
+	d := &DashboardProvisionerImpl{
+		log:         logger,
+		fileReaders: fileReaders,
 	}
 
-	for _, cfg := range cfgs {
-		switch cfg.Type {
+	return d, nil
+}
+
+func (provider *DashboardProvisionerImpl) Provision() error {
+	for _, reader := range provider.fileReaders {
+		err := reader.startWalkingDisk()
+		if err != nil {
+			return errors.Wrapf(err, "Failed to provision config %v", reader.Cfg.Name)
+		}
+	}
+
+	return nil
+}
+
+// PollChanges starts polling for changes in dashboard definition files. It creates goruotine for each provider
+// defined in the config.
+func (provider *DashboardProvisionerImpl) PollChanges(ctx context.Context) {
+	for _, reader := range provider.fileReaders {
+		go reader.pollChanges(ctx)
```

Comment:
I'm not sure we need to start goroutines for this. I usually default to not use goroutines unless we need them. 
While they are cheap, they make it trickier to debug code if there are panics.

Reply:
I do not see any way around it. The `pollChanges` function does not return until the context is cancelled as it has its own ticker it is waiting for. Each also can have different polling interval so they cannot be put into single goroutine.

Reply:
:+1: